File: frontend/types/analysis.ts
export interface LeetCodeProblem {
  title: string;
  url: string;
  difficulty: "Easy" | "Medium" | "Hard";
}

export interface CodingPattern {
  name: string;
  problems: LeetCodeProblem[];
}

export interface CodingPrep {
  patterns: CodingPattern[];
}

export interface PriorityTopic {
  topic: string;
  why: string;
  estimated_hours: number;
}

export interface PrepPlan {
  fit_score: number;
  match_summary: string;
  strengths: string[];
  skill_gaps: string[];
  interviewer_concerns: string[];
  high_priority_topics: PriorityTopic[];
  medium_priority_topics: PriorityTopic[];
  low_priority_topics: PriorityTopic[];
  what_to_ignore: string[];
  coding_prep: CodingPrep;
  system_design_topics: string[];
  behavioral_focus: string[];
  interview_signals: string[];
  glassdoor_sources?: string[];
  estimated_total_prep_days: number;
  timeline_breakdown: string;
}

export interface AnalysisStatusResponse {
  id: number;
  status: "pending" | "processing" | "completed" | "failed";
  fit_score: number | null;
  result: PrepPlan | null;
  error_message: string | null;
  created_at: string;
  company_name: string;
  role_title: string;
  from_cache: boolean;
}

export interface AnalysisListItem {
  id: number;
  status: "pending" | "processing" | "completed" | "failed";
  fit_score: number | null;
  created_at: string;
  company_name: string;
  role_title: string;
  from_cache: boolean;
}

export interface ResumeUploadResponse {
  resume_id: number;
  filename: string;
}

export interface AnalysisCreateResponse {
  analysis_id: number;
  status: string;
  from_cache: boolean;
}

export interface MyResumeResponse {
  resume_id: number;
  filename: string;
}

File: backend/app/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, Field


class LeetCodeProblem(BaseModel):
    title: str
    url: str
    difficulty: str


class CodingPattern(BaseModel):
    name: str
    problems: list[LeetCodeProblem] = Field(default_factory=list)


class CodingPrep(BaseModel):
    patterns: list[CodingPattern] = Field(default_factory=list)


class PriorityTopic(BaseModel):
    topic: str
    why: str
    estimated_hours: int


class PrepPlan(BaseModel):
    fit_score: int = Field(ge=0, le=100)
    match_summary: str
    strengths: list[str] = Field(default_factory=list)
    skill_gaps: list[str] = Field(default_factory=list)
    interviewer_concerns: list[str] = Field(default_factory=list)
    high_priority_topics: list[PriorityTopic] = Field(default_factory=list)
    medium_priority_topics: list[PriorityTopic] = Field(default_factory=list)
    low_priority_topics: list[PriorityTopic] = Field(default_factory=list)
    what_to_ignore: list[str] = Field(default_factory=list)
    coding_prep: CodingPrep = Field(default_factory=CodingPrep)
    system_design_topics: list[str] = Field(default_factory=list)
    behavioral_focus: list[str] = Field(default_factory=list)
    interview_signals: list[str] = Field(default_factory=list)
    glassdoor_sources: list[str] = Field(default_factory=list)
    estimated_total_prep_days: int = Field(ge=1)
    timeline_breakdown: str


class AnalysisCreateRequest(BaseModel):
    resume_id: int
    jd_text: str = Field(min_length=50)
    company_name: str = Field(min_length=1, max_length=200)
    role_title: str = Field(min_length=1, max_length=200)


class AnalysisStatusResponse(BaseModel):
    id: int
    status: str
    fit_score: float | None = None
    result: PrepPlan | None = None
    error_message: str | None = None
    created_at: datetime
    company_name: str
    role_title: str
    from_cache: bool = False


class AnalysisListItem(BaseModel):
    id: int
    status: str
    fit_score: float | None = None
    created_at: datetime
    company_name: str
    role_title: str
    from_cache: bool = False
